# Products Dataset Cleaning

This section prepares the **products dataset** for analysis.

The raw products table was imported into MySQL with all columns stored as `VARCHAR(100)`.  
This is acceptable for ingestion, but it is not suitable for analysis because:

- numeric columns are still stored as text
- missing values may exist
- product category names may be inconsistent
- duplicate product IDs may exist
- product records have not yet been validated for downstream joins

To address this, a new typed cleaning table called `clean_products` is created.

The cleaning process for products follows these steps:

1. Inspect the raw products table
2. Check for missing and blank values
3. Create the typed clean products table
4. Load and standardize data from the raw table
5. Inspect the cleaned data
6. Check for missing values after type conversion
7. Review product category values
8. Check for invalid numeric values
9. Replace invalid numeric values with NULL
10. Detect and remove duplicate product IDs
11. Add data quality flags
12. Validate the primary key
13. Add the primary key constraint

In [ ]:
import sys
!{sys.executable} -m pip install PyMySQL
!{sys.executable} -m pip install ipython-sql
%load_ext sql
%sql mysql+pymysql://mysql+pymysql://{DB_USER}:{DB_PASSWORD}@{DB_HOST}/{DB_NAME}
%config SqlMagic.style = 'DEFAULT'
!pip install jupysql
%config SqlMagic.displaylimit = 100

In [ ]:
pip show pymysql

# 1. Inspect the Raw Products Table

Before cleaning the data, the raw products table should be inspected.

This helps identify:

- number of records
- possible blank values
- numeric fields stored as text
- possible duplicate product IDs

In [3]:
%%sql
    
SELECT COUNT(*) AS total_rows
FROM raw_products;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

total_rows
32951


In [4]:
%%sql
    
SELECT *
FROM raw_products
LIMIT 10;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

10 rows affected.

product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40,287,1,225,16,10,14
3aa071139cb16b67ca9e5dea641aaa2f,artes,44,276,1,1000,30,18,20
96bd76ec8810374ed1b65e291975717f,esporte_lazer,46,250,1,154,18,9,15
cef67bcfe19066a932b7673e239eb23d,bebes,27,261,1,371,26,4,26
9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37,402,4,625,20,17,13
41d3672d4792049fa1779bb35283ed13,instrumentos_musicais,60,745,1,200,38,5,11
732bd381ad09e530fe0a5f457d81becb,cool_stuff,56,1272,4,18350,70,24,44
2548af3e6e77a690cf3eb6368e9ab61e,moveis_decoracao,56,184,2,900,40,8,40
37cc742be07708b53a98702e77a21a02,eletrodomesticos,57,163,1,400,27,13,17
8c92109888e8cdf9d66dc7e463025574,brinquedos,36,1156,1,600,17,10,12


# 2. Check Missing and Blank Values

This step checks whether important columns contain missing values or blank strings.

Special attention is given to:

- `product_id`
- `product_category_name`
- description and photo counts
- weight and dimension fields

In [5]:
%%sql
    
SELECT
    SUM(CASE WHEN product_id IS NULL OR TRIM(product_id) = '' THEN 1 ELSE 0 END) AS missing_product_id,
    SUM(CASE WHEN product_category_name IS NULL OR TRIM(product_category_name) = '' THEN 1 ELSE 0 END) AS missing_product_category_name,
    SUM(CASE WHEN product_name_lenght IS NULL OR TRIM(product_name_lenght) = '' THEN 1 ELSE 0 END) AS missing_product_name_lenght,
    SUM(CASE WHEN product_description_lenght IS NULL OR TRIM(product_description_lenght) = '' THEN 1 ELSE 0 END) AS missing_product_description_lenght,
    SUM(CASE WHEN product_photos_qty IS NULL OR TRIM(product_photos_qty) = '' THEN 1 ELSE 0 END) AS missing_product_photos_qty,
    SUM(CASE WHEN product_weight_g IS NULL OR TRIM(product_weight_g) = '' THEN 1 ELSE 0 END) AS missing_product_weight_g,
    SUM(CASE WHEN product_length_cm IS NULL OR TRIM(product_length_cm) = '' THEN 1 ELSE 0 END) AS missing_product_length_cm,
    SUM(CASE WHEN product_height_cm IS NULL OR TRIM(product_height_cm) = '' THEN 1 ELSE 0 END) AS missing_product_height_cm,
    SUM(CASE WHEN product_width_cm IS NULL OR TRIM(product_width_cm) = '' THEN 1 ELSE 0 END) AS missing_product_width_cm
FROM raw_products;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

missing_product_id,missing_product_category_name,missing_product_name_lenght,missing_product_description_lenght,missing_product_photos_qty,missing_product_weight_g,missing_product_length_cm,missing_product_height_cm,missing_product_width_cm
0,610,610,610,610,2,2,2,2


The `product_id` column has **no missing values**, confirming that every product record has a unique identifier. This is important because `product_id` functions as the primary key for the products table and is used in joins with other tables such as `order_items`.

- **Product Category and Text Attributes**
  
  A total of **610 records are missing values** for:
  
  - `product_category_name`
  - `product_name_length`
  - `product_description_length`
  - `product_photos_qty`
  
While these fields are useful for descriptive analysis (e.g., product catalog analysis), they **do not affect the core transaction analysis** used in the RFM segmentation because product-level descriptive attributes are not required for calculating recency, frequency, or monetary value.

- **Product Dimensions and Weight**
  
  Only **2 records are missing** values for the physical attributes:
  
  - `product_weight_g`
  - `product_length_cm`
  - `product_height_cm`
  - `product_width_cm`
  
These attributes are typically used for logistics and shipping calculations rather than customer behavior analysis. Given the very small number of missing values, they can either be left as `NULL` or handled during downstream logistics analysis if required.



# 3. Create the Typed Clean Products Table

The clean products table is created from the raw products dataset and enriched using the product category translation dataset.

This step performs several transformations to improve data usability, consistency, and analytical readiness.

The main objectives of this step are:

- convert raw text fields into appropriate data types,
- standardize text formatting,
- translate Portuguese product categories into English,
- preserve all product records for downstream analysis,
- ensure numeric fields are stored in appropriate numeric formats.

### 3.1 Product Category Translation

The raw products dataset contains category names in Portuguese.  
To improve interpretability and support English-based reporting in tools such as Power BI, the dataset is joined with the category translation table.

A **LEFT JOIN** is used to ensure that:

- all products remain in the dataset,
- translated categories are added where available,
- any unmatched categories are retained for transparency and data quality review.

This approach avoids losing product records due to missing translations.

In [7]:
%%sql
DROP TABLE IF EXISTS clean_products;

CREATE TABLE clean_products (
    product_id VARCHAR(50),
    product_category_name VARCHAR(100),
    product_category_name_english VARCHAR(100),
    product_category_final VARCHAR(100),
    product_name_length INT,
    product_description_length INT,
    product_photos_qty INT,
    product_weight_g DECIMAL(10,2),
    product_length_cm DECIMAL(10,2),
    product_height_cm DECIMAL(10,2),
    product_width_cm DECIMAL(10,2)
);

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

++
||
++
++

# 4. Load and Standardize Data from Raw Products

Data is inserted from `raw_products` into `clean_products`.

During this step:

- `TRIM()` == removes extra spaces
- `NULLIF(..., '')` == converts blank strings into NULL
- integer-like fields are cast from text to numeric form
- weight and dimension fields are cast to decimal values
- `product_category_name` is standardized to lowercase

In [8]:
%%sql
INSERT INTO clean_products (
    product_id,
    product_category_name,
    product_category_name_english,
    product_category_final,
    product_name_length,
    product_description_length,
    product_photos_qty,
    product_weight_g,
    product_length_cm,
    product_height_cm,
    product_width_cm
)
SELECT
    NULLIF(TRIM(p.product_id), '') AS product_id,
    LOWER(NULLIF(TRIM(p.product_category_name), '')) AS product_category_name,
    LOWER(NULLIF(TRIM(t.product_category_name_english), '')) AS product_category_name_english,
    COALESCE(
        LOWER(NULLIF(TRIM(t.product_category_name_english), '')),
        LOWER(NULLIF(TRIM(p.product_category_name), ''))
    ) AS product_category_final,
    CAST(NULLIF(TRIM(p.product_name_lenght), '') AS UNSIGNED) AS product_name_length,
    CAST(NULLIF(TRIM(p.product_description_lenght), '') AS UNSIGNED) AS product_description_length,
    CAST(NULLIF(TRIM(p.product_photos_qty), '') AS UNSIGNED) AS product_photos_qty,
    CAST(NULLIF(TRIM(p.product_weight_g), '') AS DECIMAL(10,2)) AS product_weight_g,
    CAST(NULLIF(TRIM(p.product_length_cm), '') AS DECIMAL(10,2)) AS product_length_cm,
    CAST(NULLIF(TRIM(p.product_height_cm), '') AS DECIMAL(10,2)) AS product_height_cm,
    CAST(NULLIF(TRIM(p.product_width_cm), '') AS DECIMAL(10,2)) AS product_width_cm
FROM raw_products p
LEFT JOIN raw_product_category_name_translation t
    ON LOWER(TRIM(p.product_category_name)) = LOWER(TRIM(t.product_category_name));

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

32951 rows affected.

++
||
++
++

In [9]:
%%sql
    
SELECT *
FROM clean_products
LIMIT 10;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

10 rows affected.

product_id,product_category_name,product_category_name_english,product_category_final,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,perfumery,perfumery,40,287,1,225.00,16.00,10.00,14.00
3aa071139cb16b67ca9e5dea641aaa2f,artes,art,art,44,276,1,1000.00,30.00,18.00,20.00
96bd76ec8810374ed1b65e291975717f,esporte_lazer,sports_leisure,sports_leisure,46,250,1,154.00,18.00,9.00,15.00
cef67bcfe19066a932b7673e239eb23d,bebes,baby,baby,27,261,1,371.00,26.00,4.00,26.00
9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,housewares,housewares,37,402,4,625.00,20.00,17.00,13.00
41d3672d4792049fa1779bb35283ed13,instrumentos_musicais,musical_instruments,musical_instruments,60,745,1,200.00,38.00,5.00,11.00
732bd381ad09e530fe0a5f457d81becb,cool_stuff,cool_stuff,cool_stuff,56,1272,4,18350.00,70.00,24.00,44.00
2548af3e6e77a690cf3eb6368e9ab61e,moveis_decoracao,furniture_decor,furniture_decor,56,184,2,900.00,40.00,8.00,40.00
37cc742be07708b53a98702e77a21a02,eletrodomesticos,home_appliances,home_appliances,57,163,1,400.00,27.00,13.00,17.00
8c92109888e8cdf9d66dc7e463025574,brinquedos,toys,toys,36,1156,1,600.00,17.00,10.00,12.00


# 5. Check Missing Values in Clean Products

After type conversion and standardization, the clean table should be checked again for missing values in important fields.

In [11]:
%%sql
    
SELECT
    SUM(product_id IS NULL) AS missing_product_id,
    SUM(product_category_name IS NULL) AS missing_product_category_name,
    SUM(product_category_name_english IS NULL) AS missing_product_category_name_english,
    SUM(product_category_final IS NULL) AS missing_product_category_name_final,
    SUM(product_name_length IS NULL) AS missing_product_name_length,
    SUM(product_description_length IS NULL) AS missing_product_description_length,
    SUM(product_photos_qty IS NULL) AS missing_product_photos_qty,
    SUM(product_weight_g IS NULL) AS missing_product_weight_g,
    SUM(product_length_cm IS NULL) AS missing_product_length_cm,
    SUM(product_height_cm IS NULL) AS missing_product_height_cm,
    SUM(product_width_cm IS NULL) AS missing_product_width_cm
FROM clean_products;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

missing_product_id,missing_product_category_name,missing_product_category_name_english,missing_product_category_name_final,missing_product_name_length,missing_product_description_length,missing_product_photos_qty,missing_product_weight_g,missing_product_length_cm,missing_product_height_cm,missing_product_width_cm
0,610,623,610,610,610,610,2,2,2,2


In [12]:
%%sql
    
SELECT DISTINCT
    product_category_name
FROM clean_products
WHERE product_category_name IS NOT NULL
  AND product_category_name_english IS NULL
ORDER BY product_category_name;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

2 rows affected.

product_category_name
pc_gamer
portateis_cozinha_e_preparadores_de_alimentos


### Handling Missing Product Category Translations

During the product category translation process, several categories in the products dataset did not have corresponding entries in the translation lookup table.

To maintain consistency and completeness of the dataset, the missing category translations were manually added to the translation table.

This ensures that:

- all product categories have a corresponding English label,
- category naming remains standardized across the dataset,
- the translation mapping remains centralized in the lookup table.

Manual additions were limited only to categories that were confirmed to be missing from the translation dataset.

After inserting the missing translations, the clean products table was regenerated so that the translated categories would be populated correctly during the join process.

This method preserves transparency and reproducibility in the data cleaning workflow.

In [16]:
%%sql
    
INSERT INTO raw_product_category_name_translation
(product_category_name, product_category_name_english)
VALUES
('portateis_cozinha_e_preparadores_de_alimentos', 'Small kitchen appliances and food processors/preparers')
;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

++
||
++
++

In [17]:
%%sql
SELECT *
FROM raw_product_category_name_translation
WHERE product_category_name = 'portateis_cozinha_e_preparadores_de_alimentos';

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

product_category_name,product_category_name_english
portateis_cozinha_e_preparadores_de_alimentos,Small kitchen appliances and food processors/preparers


In [19]:
%%sql
TRUNCATE TABLE clean_products;

INSERT INTO clean_products (
    product_id,
    product_category_name,
    product_category_name_english,
    product_category_final,
    product_name_length,
    product_description_length,
    product_photos_qty,
    product_weight_g,
    product_length_cm,
    product_height_cm,
    product_width_cm
)
SELECT
    NULLIF(TRIM(p.product_id), '') AS product_id,
    LOWER(NULLIF(TRIM(p.product_category_name), '')) AS product_category_name,
    LOWER(NULLIF(TRIM(t.product_category_name_english), '')) AS product_category_name_english,
    COALESCE(
        LOWER(NULLIF(TRIM(t.product_category_name_english), '')),
        LOWER(NULLIF(TRIM(p.product_category_name), ''))
    ) AS product_category_final,
    CAST(NULLIF(TRIM(p.product_name_lenght), '') AS UNSIGNED) AS product_name_length,
    CAST(NULLIF(TRIM(p.product_description_lenght), '') AS UNSIGNED) AS product_description_length,
    CAST(NULLIF(TRIM(p.product_photos_qty), '') AS UNSIGNED) AS product_photos_qty,
    CAST(NULLIF(TRIM(p.product_weight_g), '') AS DECIMAL(10,2)) AS product_weight_g,
    CAST(NULLIF(TRIM(p.product_length_cm), '') AS DECIMAL(10,2)) AS product_length_cm,
    CAST(NULLIF(TRIM(p.product_height_cm), '') AS DECIMAL(10,2)) AS product_height_cm,
    CAST(NULLIF(TRIM(p.product_width_cm), '') AS DECIMAL(10,2)) AS product_width_cm
FROM raw_products p
LEFT JOIN raw_product_category_name_translation t
    ON LOWER(TRIM(p.product_category_name)) = LOWER(TRIM(t.product_category_name));

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

32951 rows affected.

++
||
++
++

# 6. Check Missing Values in Clean Products

After type conversion and standardization, the clean table should be checked again for missing values in important fields.

In [20]:
%%sql
    
SELECT
    SUM(product_id IS NULL) AS missing_product_id,
    SUM(product_category_name IS NULL) AS missing_product_category_name,
    SUM(product_category_name_english IS NULL) AS missing_product_category_name_english,
    SUM(product_category_final IS NULL) AS missing_product_category_name_final,
    SUM(product_name_length IS NULL) AS missing_product_name_length,
    SUM(product_description_length IS NULL) AS missing_product_description_length,
    SUM(product_photos_qty IS NULL) AS missing_product_photos_qty,
    SUM(product_weight_g IS NULL) AS missing_product_weight_g,
    SUM(product_length_cm IS NULL) AS missing_product_length_cm,
    SUM(product_height_cm IS NULL) AS missing_product_height_cm,
    SUM(product_width_cm IS NULL) AS missing_product_width_cm
FROM clean_products;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

missing_product_id,missing_product_category_name,missing_product_category_name_english,missing_product_category_name_final,missing_product_name_length,missing_product_description_length,missing_product_photos_qty,missing_product_weight_g,missing_product_length_cm,missing_product_height_cm,missing_product_width_cm
0,610,613,610,610,610,610,2,2,2,2


The `product_id` column has **no missing values**, confirming that each product record contains a valid identifier. This is important because `product_id` acts as the primary key in the products table and is used to join with the `order_items` table.

**Product Category Fields**

The fields `product_category_name` and `product_category_name_final` contain **610 missing values**, indicating that some products in the raw dataset do not have an associated category. These missing values propagate to the final category column because the final category is derived from the original category and its English translation.

The `product_category_name_english` column shows **623 missing values**, which is slightly higher than the missing original category count. This indicates that some categories exist in the raw dataset but **do not have a matching translation in the category translation lookup table**.

The missing values identified in the products dataset are **not expected to impact the RFM segmentation analysis**, as RFM calculations rely on:

- order purchase timestamps
- order frequency
- payment values

However, product attributes and categories remain important for **product-level analysis and category-based revenue insights** in Power BI dashboards. Therefore, the missing values are documented and retained to maintain dataset transparency and completeness.

# 7. Inspect the Clean Products Table

After loading the data into the typed clean table, inspect a sample of rows to verify that:

- IDs loaded correctly
- numeric fields converted properly
- category standardization worked as expected
- null handling worked as expected

In [21]:
%%sql
    
SELECT *
FROM clean_products
LIMIT 10;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

10 rows affected.

product_id,product_category_name,product_category_name_english,product_category_final,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,perfumery,perfumery,40,287,1,225.00,16.00,10.00,14.00
3aa071139cb16b67ca9e5dea641aaa2f,artes,art,art,44,276,1,1000.00,30.00,18.00,20.00
96bd76ec8810374ed1b65e291975717f,esporte_lazer,sports_leisure,sports_leisure,46,250,1,154.00,18.00,9.00,15.00
cef67bcfe19066a932b7673e239eb23d,bebes,baby,baby,27,261,1,371.00,26.00,4.00,26.00
9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,housewares,housewares,37,402,4,625.00,20.00,17.00,13.00
41d3672d4792049fa1779bb35283ed13,instrumentos_musicais,musical_instruments,musical_instruments,60,745,1,200.00,38.00,5.00,11.00
732bd381ad09e530fe0a5f457d81becb,cool_stuff,cool_stuff,cool_stuff,56,1272,4,18350.00,70.00,24.00,44.00
2548af3e6e77a690cf3eb6368e9ab61e,moveis_decoracao,furniture_decor,furniture_decor,56,184,2,900.00,40.00,8.00,40.00
37cc742be07708b53a98702e77a21a02,eletrodomesticos,home_appliances,home_appliances,57,163,1,400.00,27.00,13.00,17.00
8c92109888e8cdf9d66dc7e463025574,brinquedos,toys,toys,36,1156,1,600.00,17.00,10.00,12.00


# 8. Review Distinct Product Categories

Product category names should be standardized and reviewed before downstream analysis.

This step helps identify inconsistent category values and missing categories.

In [27]:
%%sql
    
SELECT DISTINCT product_category_name, product_category_name_english
FROM clean_products
ORDER BY product_category_name

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

74 rows affected.

product_category_name,product_category_name_english
None,None
agro_industria_e_comercio,agro_industry_and_commerce
alimentos,food
alimentos_bebidas,food_drink
artes,art
artes_e_artesanato,arts_and_craftmanship
artigos_de_festas,party_supplies
artigos_de_natal,christmas_supplies
audio,audio
automotivo,auto


# 9.  Check for Invalid Numeric Values

The following fields should not contain negative values:

- `product_name_lenght`
- `product_description_lenght`
- `product_photos_qty`
- `product_weight_g`
- `product_length_cm`
- `product_height_cm`
- `product_width_cm`

These values are checked before analysis.

In [32]:
%%sql
    
SELECT *
FROM clean_products
WHERE product_name_length < 0
   OR product_description_length < 0
   OR product_photos_qty < 0
   OR product_weight_g < 0
   OR product_length_cm < 0
   OR product_height_cm < 0
   OR product_width_cm < 0;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

product_id,product_category_name,product_category_name_english,product_category_final,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm


# 10.  Replace Invalid Numeric Values with NULL

Instead of inventing replacement values, clearly invalid numeric values are converted to NULL.

In [33]:
%%sql
    
SELECT
    product_id,
    COUNT(*) AS duplicate_count
FROM clean_products
GROUP BY product_id
HAVING COUNT(*) > 1;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

product_id,duplicate_count


# 11. Add Data Quality Flags

Instead of dropping every imperfect row, data quality flags help preserve records while indicating where issues exist.

In this step, two useful flags are added:

- `missing_category_flag`
- `missing_dimension_flag`

In [34]:
%%sql
    
ALTER TABLE clean_products
ADD COLUMN missing_category_flag TINYINT DEFAULT 0,
ADD COLUMN missing_dimension_flag TINYINT DEFAULT 0;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

++
||
++
++

In [35]:
%%sql
    
UPDATE clean_products
SET
    missing_category_flag = CASE
        WHEN product_category_name IS NULL THEN 1
        ELSE 0
    END,
    missing_dimension_flag = CASE
        WHEN product_weight_g IS NULL
          OR product_length_cm IS NULL
          OR product_height_cm IS NULL
          OR product_width_cm IS NULL
        THEN 1
        ELSE 0
    END;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

32951 rows affected.

++
||
++
++

# 12. Validate the Product Primary Key

This step confirms that the `product_id` column is now:

- unique
- non-null
- suitable to be used as the primary key

In [36]:
%%sql
    
SELECT
    COUNT(*) AS total_rows,
    COUNT(product_id) AS non_null_ids,
    COUNT(DISTINCT product_id) AS distinct_ids
FROM clean_products;

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

1 rows affected.

total_rows,non_null_ids,distinct_ids
32951,32951,32951


# 13. Add the Products Primary Key

After confirming the uniqueness and validity of `product_id`, the primary key constraint can be added.

In [38]:
%%sql
    
ALTER TABLE clean_products
MODIFY product_id VARCHAR(50) NOT NULL,
ADD PRIMARY KEY (product_id);

Running query in 'mysql+pymysql://root:***@localhost/retail_customer_segmentation'

RuntimeError: (pymysql.err.OperationalError) (1068, 'Multiple primary key defined')
[SQL: ALTER TABLE clean_products
MODIFY product_id VARCHAR(50) NOT NULL,
ADD PRIMARY KEY (product_id);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


# Products Dataset Cleaning Complete

The products dataset has now been:

- converted from text to correct data types
- standardized for blank and null values
- standardized for category text values
- checked for invalid numeric values
- validated for duplicate product IDs
- given data quality flags
- validated for primary key integrity

The `clean_products` table is now ready for downstream processes such as:

- linking to order items
- product category analysis
- basket-level analysis
- product-level revenue aggregation
- product enrichment for Power BI dashboards

# Products Dataset Cleaning Complete

The products dataset has now been:

- converted from text to correct data types
- standardized for blank and null values
- standardized for category text values
- checked for invalid numeric values
- validated for duplicate product IDs
- given data quality flags
- validated for primary key integrity

The `clean_products` table is now ready for downstream processes such as:

- linking to order items
- product category analysis
- basket-level analysis
- product-level revenue aggregation
- product enrichment for Power BI dashboards